File: languages/jvm/java.com/helidon.io/src/main/java/com/proxy/reverse/Application.java
package com.proxy.reverse;

import io.helidon.webclient.api.HttpClientResponse;
import io.helidon.webclient.api.WebClient;
import io.helidon.webserver.WebServer;
import io.helidon.webserver.http.ServerRequest;
import io.helidon.webserver.http.ServerResponse;
import java.io.InputStream;
import java.io.OutputStream;

public class Application {
  public static void main(String[] args) {
    WebClient client = WebClient.builder().build();

    WebServer server =
        WebServer.builder()
            .port(8080)
            .routing(r -> r.get("/proxy", (req, res) -> handleProxy(req, res, client)))
            .build()
            .start();

    System.out.println("✅ Helidon Reverse Proxy running at http://localhost:" + server.port());
  }

  private static void handleProxy(ServerRequest req, ServerResponse res, WebClient client) {
    String targetUrl = req.query().first("url").orElse(null);

    if (targetUrl == null || targetUrl.isBlank()) {
      res.status(400).send("Missing 'url' query parameter");
      return;
    }

    try {
      HttpClientResponse response = client.get(targetUrl).request();
      InputStream input = response.inputStream();
      OutputStream output = res.outputStream();
      input.transferTo(output);
    } catch (Exception e) {
      res.status(500).send(e.getMessage());
    }
  }
}

File: languages/jvm/kotlinlang.org/ktor/src/main/kotlin/com/proxy/reverse/Application.kt
package com.proxy.reverse

import io.ktor.server.application.*
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import io.ktor.server.engine.*
import io.ktor.server.netty.*
import kotlinx.coroutines.*

fun main() {
  embeddedServer(Netty, port = 8080, host = "0.0.0.0") {
    module()
  }.start(wait = true)
}

fun Application.module() {
  val client = HttpClient(CIO)

  routing {
    route("/api") {
      handle {
        val targetUrl = call.request.queryParameters["url"]
          ?: return@handle call.respond(HttpStatusCode.BadRequest, "Missing 'url' query parameter")

        val method = call.request.httpMethod
        val requestHeaders = call.request.headers

        val proxiedResponse: HttpResponse = when (method) {
          HttpMethod.Get -> client.get(targetUrl) {
            headers {
              requestHeaders.forEach { key, values -> appendAll(key, values) }
            }
          }
          HttpMethod.Post -> {
            val requestBody = call.receiveText()
            client.post(targetUrl) {
              headers {
                requestHeaders.forEach { key, values -> appendAll(key, values) }
              }
              setBody(requestBody)
            }
          }
          else -> return@handle call.respond(HttpStatusCode.MethodNotAllowed, "Only GET and POST supported")
        }

        val responseBytes = proxiedResponse.body<ByteArray>()
        call.respondBytes(
          bytes = responseBytes,
          contentType = proxiedResponse.contentType(),
          status = proxiedResponse.status
        )
      }
    }
  }
}
